Dockerfile

```diff
@@ -9,10 +9,23 @@ RUN     curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
 
 WORKDIR /meilisearch
 
-COPY    . .
+COPY    Cargo.lock .
+COPY    Cargo.toml .
+
+COPY    meilisearch-error/Cargo.toml meilisearch-error/
+COPY    meilisearch-http/Cargo.lock meilisearch-http/
+COPY    meilisearch-http/Cargo.toml meilisearch-http/
 
 ENV     RUSTFLAGS="-C target-feature=-crt-static"
 
+# Create dummy main.rs files for each workspace member to be able to compile all the dependencies
+RUN     find . -type d -name "meilisearch-*" | xargs -I{} sh -c 'mkdir {}/src; echo "fn main() { }" > {}/src/main.rs;'
+# Use `cargo build` instead of `cargo vendor` because we need to not only download but compile dependencies too
+RUN     $HOME/.cargo/bin/cargo build --release
+# Cleanup dummy main.rs files
+RUN     find . -path "*/src/main.rs" -delete
```

Comment:
I am not sure I am right but there already is a main.rs file in the `meilisearch-http` crate, also won't the `cargo build --release` work if there is a lib.rs file in the crate (like for `meilisearch-error`)?

Reply:
yes it is, the thing is that I'm not copying the actual `meilisearch-http/src/main.rs` at the dependencies build step. Instead I create a dummy one to be able to build dependencies for all the workspace members and only after that copy actual sources such as `meilisearch-http/src/main.rs`
https://github.com/meilisearch/transplant/blob/5bc464dc53db14e222a955b2e9ae1960089b4d84/Dockerfile#L28

Reply:
Yeah, I saw that you were erasing the dummy main.rs files afterward.

Have you tested that the server was running? I am sure that in Rust, when it compiles it works but... you now? :)

Reply:
regarding the `meilisearch-error/src/lib.rs: we don't actually need to fully compile the source code of all workspace members, instead we want only dependencies they use. So if we'll copy actual lib.rs, it most likely won't compile because it will need other source files it imports. And if we import them too, we'll compile the whole `meilisearch-error/src` as before 😄

Reply:
Ho yeah, sorry, you were only copying the Cargo.toml and Cargo.lock files, creating a dummy main.rs file, just to let cargo fetch and compile the MeiliSearch dependencies! Sorry, I understand now!

Reply:
> Have you tested that the server was running?

yep
```console
➜ transplant (docker-deps-cache) ✔ docker build -t meilisearch .
Sending build context to Docker daemon  864.3kB
Step 1/24 : FROM    alpine:3.10 AS compiler
 ---> d1bb6234ef26
Step 2/24 : RUN     apk update --quiet
 ---> Using cache
 ---> 5a5a2563623b
Step 3/24 : RUN     apk add curl
 ---> Using cache
 ---> 6227ce30d4c9
Step 4/24 : RUN     apk add build-base
 ---> Using cache
 ---> d4bf14b3868e
Step 5/24 : RUN     curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
 ---> Using cache
 ---> a7288a8a328e
Step 6/24 : WORKDIR /meilisearch
 ---> Using cache
 ---> b11e8c9b7975
Step 7/24 : COPY    Cargo.lock .
 ---> Using cache
 ---> 848c78a7c116
Step 8/24 : COPY    Cargo.toml .
 ---> Using cache
 ---> 6516eb971761
Step 9/24 : COPY    meilisearch-error/Cargo.toml meilisearch-error/
 ---> Using cache
 ---> 35e83b568984
Step 10/24 : COPY    meilisearch-http/Cargo.lock meilisearch-http/
 ---> Using cache
 ---> c9f598266bce
Step 11/24 : COPY    meilisearch-http/Cargo.toml meilisearch-http/
 ---> Using cache
 ---> 0756fdeed181
Step 12/24 : ENV     RUSTFLAGS="-C target-feature=-crt-static"
 ---> Using cache
 ---> bb7b67540cd6
Step 13/24 : RUN     find . -type d -name "meilisearch-*" | xargs -I{} sh -c 'mkdir {}/src; echo "fn main() { }" > {}/src/main.rs;'
 ---> Using cache
 ---> bcfbc6f22ba5
Step 14/24 : RUN     $HOME/.cargo/bin/cargo build --release
 ---> Using cache
 ---> 5e99db3abdb0
Step 15/24 : RUN     find . -path "*/src/main.rs" -delete
 ---> Using cache
 ---> 0ca82cfe545e
Step 16/24 : COPY    . .
 ---> 99fca3061bbd
Step 17/24 : RUN     $HOME/.cargo/bin/cargo build --release
 ---> Running in 708c8e9e2d03
   Compiling serde v1.0.124
   Compiling bitflags v1.2.1
   Compiling num-traits v0.2.14
   Compiling num-integer v0.1.44
   Compiling semver-parser v0.10.2
   Compiling time v0.1.44
   Compiling semver v0.11.0
   Compiling rustc_version v0.3.3
   Compiling meilisearch-error v0.19.0 (/meilisearch/meilisearch-error)
   Compiling chrono v0.4.19
   Compiling vergen v3.2.0
   Compiling meilisearch-http v0.21.0-alpha.1 (/meilisearch/meilisearch-http)
    Finished release [optimized + debuginfo] target(s) in 3m 43s
Removing intermediate container 708c8e9e2d03
 ---> 5bda902c9ba9
Step 18/24 : FROM    alpine:3.10
 ---> d1bb6234ef26
Step 19/24 : RUN     apk add -q --no-cache libgcc tini
 ---> Using cache
 ---> b709b4e9f253
Step 20/24 : COPY    --from=compiler /meilisearch/target/release/meilisearch .
 ---> 50577693aa8a
Step 21/24 : ENV     MEILI_HTTP_ADDR 0.0.0.0:7700
 ---> Running in a6b84a7cdca1
Removing intermediate container a6b84a7cdca1
 ---> e11042a4abaa
Step 22/24 : EXPOSE  7700/tcp
 ---> Running in c691408d0196
Removing intermediate container c691408d0196
 ---> 0c2c68a23fe1
Step 23/24 : ENTRYPOINT ["tini", "--"]
 ---> Running in fc857e9c98f0
Removing intermediate container fc857e9c98f0
 ---> de6f4f4f39c4
Step 24/24 : CMD     ./meilisearch
 ---> Running in 04aecc496a4f
Removing intermediate container 04aecc496a4f
 ---> e5268ecc7ccf
Successfully built e5268ecc7ccf
Successfully tagged meilisearch:latest

➜ transplant (docker-deps-cache) ✔ docker run -it -d -p 7700:7700 meilisearch ./meilisearch --no-analytics=true
aed0316b90bd7c432e38732a2f4e9218c054fc83354683166f052d04af4490d4

➜ transplant (docker-deps-cache) ✔ curl http://localhost:7700/version
{"commitSha":"UNKNOWN","buildDate":"2021-03-31T10:19:28.507821800+00:00","pkgVersion":"0.21.0-alpha.1"}
```

Reply:
I tested to build and run the image, it works 🙂